README.md
<pre>
OOOOOOO 	OOOOOO  	OOOOOO  	OOO     	        	        	        	
 OO   O 	O OO O  	 OO  OO 	 OO     	        	        	        	
 OO O   	  OO    	 OO  OO 	 OO OO  	 OOOO   	OOOOO   	 OOOO   	
 OOOO   	  OO    	 OOOOO  	 OOO OO 	OO  OO  	OO  OO  	OO  OO  	
 OO O   	  OO    	 OO     	 OO  OO 	OO  OO  	OO  OO  	OOOOOO  	
 OO     	  OO    	 OO     	 OO  OO 	OO  OO  	OO  OO  	OO      	
OOOO    	 OOOO   	OOOO    	OOO  OO 	 OOOO   	OO  OO  	 OOOO   	ver. 0.1
</pre>

FTPhone - It's simple console application. It tries to emulate walky-talky. 
If you have an access to ftp server and read/write permission for 
any remote directory on this server, and this same permission har your friend,
you can try to talk to him using this program ;-)

It's very easy. 
[1] Compile program:
	./make.sh

[2] And if everything OK start program with correct parameters and that all ;)
	
	sh ftphone.sh [-m my user name] [-f friend user name] [-s ftp server address] [-d remote dir]

default values:
	my user name on ftp server: kuba
	friend user name on ftp server: john
	ftp server address: ftp.server.net
	remote dir, where we will share files : /tmp/


eg.:.	sh ftphone.sh  -m kuba -f john -s ftp.server.net -d /tmp/



ChangeLog
2004-02-25 23:45  kuba

	* Src/Base.hxx: [no log message]

2004-02-25 23:45  kuba

	* Src/Base.cxx: TransferFile is virtual method now. Now it's easy
	  to change transfer protocol for gsm files.

2004-02-17 00:51  kuba

	* AUTHORS: [no log message]

2004-02-14 18:59  kuba

	* Doc/devel-doc/html/: CVS_2Base_2Base_8cxx.html,
	  CVS_2Base_2Base_8hxx-source.html, CVS_2Base_2Base_8hxx.html,
	  CVS_2Base_2Debug_8hxx-source.html, CVS_2Base_2Debug_8hxx.html,
	  CVS_2Base_2defaults_8h-source.html, CVS_2Base_2defaults_8h.html,
	  Makefile.html, ginger_8c.html, gsm-1_80-pl10_2Makefile.html,
	  sour_8c.html: devel documentation

2004-02-10 02:03  kuba

	* Src/: dsp.h, dsp.h_5.2-RELEASE.patch: [no log message]

2004-02-09 22:08  kuba

	* History/ARCHIVE_DIR, Log/DEBUG_DIR: [no log message]

2004-02-09 22:04  kuba

	* Doc/devel-doc/html/: CVS_2Base_2Base_8hxx-source.html,
	  CVS_2Base_2Base_8hxx.html, CVS_2Base_2Debug_8hxx-source.html,
	  CVS_2Base_2Debug_8hxx.html, CVS_2Base_2defaults_8h-source.html,
	  CVS_2Base_2defaults_8h.html: devel documentation

2004-02-09 22:04  kuba

	* Doc/devel-doc/html/: CVS_2Base_2Base_8cxx.html, Makefile.html,
	  ginger_8c.html, gsm-1_80-pl10_2Makefile.html, sour_8c.html: [no
	  log message]

2004-02-09 01:25  kuba

	* AUTHORS, ChangeLog: [no log message]

2004-02-09 01:11  kuba

	* Src/: Base.cxx, Base.hxx, Debug.hxx, defaults.h, dsp.c,
	  Interface.cxx, Interface.hxx, SysTools.cxx, SysTools.hxx: [no log
	  message]

2004-02-09 01:09  kuba

	* Doc/devel-doc/: foot.html, head.html, style.css,
	  html/CVS_2Base_2Base_8cxx.html, html/Makefile.html,
	  html/ginger_8c.html, html/gsm-1_80-pl10_2Makefile.html,
	  html/sour_8c.html:  user documentation

2004-02-09 00:42  kuba

	* Doc/user-doc/main.html:  user documentation

2004-02-09 00:41  kuba

	* Doc/user-doc/: index.html, FTPhone.png, valid-html401.png,
	  vcss.png: [no log message]

2004-02-05 01:16  kuba

	* Doc/user-doc/index.html: [no log message]

2004-02-04 01:17  kuba

	* Doc/user-doc/index.html:  user documentation

2004-02-04 01:12  kuba

	* Doc/user-doc/: FTPhone.png, screenshot1.png, screenshot2.png,
	  screenshot3.png, valid-html401.png, vcss.png: i've added some
	  images for user documentation

2004-01-31 19:44  kuba

	* AUTHORS, HEADER, README, Src/Debug.hxx, Src/Doxyfile: [no log
	  message]

2004-01-31 16:32  kuba

	* Src/: Main.cxx, Makefile, SysTools.cxx, SysTools.hxx: [no log
	  message]

2004-01-31 16:32  kuba

	* Src/Interface.hxx: new properties

2004-01-31 16:32  kuba

	* Src/Interface.cxx: Implementation function SetAction which inform
	  user about actual application's action.

	  Two new properties getRecWin and getPlayWin which return pointer
	  to WINDOWS structure.

2004-01-31 16:30  kuba

	* Src/Base.hxx: comments changes ;)

2004-01-31 16:29  kuba

	* Src/Base.cxx: using interface function -> SetAction

2004-01-19 20:41  kuba

	* Src/Base.cxx, ChangeLog: [no log message]

2004-01-17 17:04  kuba

	* ChangeLog: [no log message]

2004-01-17 16:59  kuba

	* Src/SysTools.cxx: [no log message]

2004-01-17 16:55  kuba

	* Src/Base.hxx: move some macros

2004-01-17 16:55  kuba

	* Src/Base.cxx: changed process of puting file on FTP server (after
	  successful uploading rename file)

2004-01-15 00:13  kuba

	* ChangeLog: [no log message]

2004-01-15 00:11  kuba

	* Src/: Base.cxx, Base.hxx: correct *.gsm format files changed
	  sleep-time in TransferFile method

2004-01-12 01:41  kuba

	* Src/Base.cxx: little change about gsm copression

2004-01-06 02:41  kuba

	* Src/dsp.c: [no log message]

2004-01-06 02:37  kuba

	* ChangeLog: [no log message]

2004-01-06 02:31  kuba

	* Src/: Interface.cxx, wave.c: [no log message]

2004-01-06 02:30  kuba

	* Src/wave.c: changes in 57, 78 line WAV_MONO -> WAV_STEREO

2004-01-06 02:28  kuba

	* Src/wave.c: some my changes for STEREO supporting

2004-01-06 02:20  kuba

	* Src/ftplib.h: for better compiling g++

2004-01-06 02:18  kuba

	* Src/dsp.c: [no log message]

2004-01-06 02:16  kuba

	* Src/Makefile: new path to gsm

2004-01-06 02:15  kuba

	* Src/Base.hxx: [no log message]

2004-01-06 02:13  kuba

	* Src/: Base.cxx, dsp.c: new gsm lib

2004-01-06 02:09  kuba

	* Src/Base.hxx: [no log message]

2004-01-06 02:07  kuba

	* Src/Base.cxx: changes for 2 channel recordnig (i suppose it's
	  failed ;( )

2004-01-06 01:17  kuba

	* ChangeLog, Src/Base.cxx, Src/Makefile, Src/dsp.c, Src/dsp.h: [no
	  log message]

2004-01-06 01:16  kuba

	* Src/gsm-1.0-pl10/COPYRIGHT, Src/gsm-1.0-pl10/ChangeLog,
	  Src/gsm-1.0-pl10/INSTALL, Src/gsm-1.0-pl10/MACHINES,
	  Src/gsm-1.0-pl10/MANIFEST, Src/gsm-1.0-pl10/Makefile,
	  Src/gsm-1.0-pl10/README, Src/gsm-1.0-pl10/add-test/add_test.c,
	  Src/gsm-1.0-pl10/add-test/add_test.dta,
	  Doc/devel-doc/html/Base_8cxx-example.html,
	  Doc/devel-doc/html/Base_8cxx.html,
	  Doc/devel-doc/html/Base_8hxx-source.html,
	  Doc/devel-doc/html/Base_8hxx.html,
	  Doc/devel-doc/html/Debug_8hxx-source.html,
	  Doc/devel-doc/html/Debug_8hxx.html,
	  Doc/devel-doc/html/Interface_8cxx.html,
	  Doc/devel-doc/html/Interface_8hxx-source.html,
	  Doc/devel-doc/html/Interface_8hxx.html,
	  Doc/devel-doc/html/Main_8cxx.html,
	  Doc/devel-doc/html/SysTools_8cxx.html,
	  Doc/devel-doc/html/SysTools_8hxx-source.html,
	  Doc/devel-doc/html/SysTools_8hxx.html,
	  Doc/devel-doc/html/add_8c.html,
	  Doc/devel-doc/html/add__test_8c.html,
	  Doc/devel-doc/html/annotated.html,
	  Doc/devel-doc/html/bitter_8c.html,
	  Doc/devel-doc/html/classTBase.html,
	  Doc/devel-doc/html/classTInterface.html,
	  Doc/devel-doc/html/cod2lin_8c.html,
	  Doc/devel-doc/html/cod2txt_8c.html,
	  Doc/devel-doc/html/code_8c.html,
	  Doc/devel-doc/html/config_8h-source.html,
	  Doc/devel-doc/html/config_8h.html,
	  Doc/devel-doc/html/debug_8c.html,
	  Doc/devel-doc/html/decode_8c.html,
	  Doc/devel-doc/html/defaults_8h-source.html,
	  Doc/devel-doc/html/defaults_8h.html,
	  Doc/devel-doc/html/doxygen.css, Doc/devel-doc/html/doxygen.png,
	  Doc/devel-doc/html/dsp_8c.html,

Src/wave.c
/* $Header: /home/cvs/FTPhone/Src/wave.c,v 1.6 2004/01/06 01:31:54 kuba Exp $ 
 * $Log: wave.c,v $
 * Revision 1.6  2004/01/06 01:31:54  kuba
 * *** empty log message ***
 *
 * Revision 1.5  2004/01/06 01:30:14  kuba
 * changes in 57, 78 line WAV_MONO -> WAV_STEREO
 *
 */

#include "wave.h"


#define MAXFILES 32

int verbose=0;
char *PRG=0;

FILE *openfile[MAXFILES];
int openfiles = 0;


const char *ERRMSG[ERRMSGS] =
{
  "Terminated without errors",
  "Terminated due to previous errors",
  "Invalid time specification",
  "Invalid frequency specification",
  "Invalid parameter",
  "Invalid Wave-File",
  "Unsupported Wave-File-Format",
  "Read Error",
  "Write Error",
  "Out of Memory",
  "Invalid syntax",
  "Too many arguments"
};


void readheader(wavfile *w)
{
  LONG l;
  LONG lriff=0,llist=0;
  /*   LONG lfmt=0; */
  /*   LONG ldata=0; */
  LONG param[7];
  
  fgetlong(&l,w->f); 
  if(l!=WAV_RIFF) errexit(ERR_FORMAT);
  fgetlong(&lriff,w->f);
  fgetlong(&l,w->f);
  if(l!=WAV_WAVE) errexit(ERR_FORMAT);
  fgetlong(&l,w->f);
  if(l==WAV_LIST)
  {
    fgetlong(&llist,w->f);
    if(fseek(w->f,llist,1)) errexit(ERR_READ);
    fgetlong(&l,w->f);
  };
  if(l!=WAV_FMT) errexit(ERR_FORMAT);
  if(fread(param,4,7,w->f)!=7 || ferror(w->f)) errexit(ERR_READ);
  if(param[0]!=16 || param[5]!=WAV_DATA) errexit(ERR_FORMAT);
  /* if(param[1]!=WAV_MONO) errexit(ERR_UNSUPP); */
  w->rate=param[2];
  w->bytes=(param[4] & 0xffff);
  if(w->bytes>sizeof(LONGINT)) errexit(ERR_UNSUPP);
  w->bits=(param[4] >> 16);
  w->n=(param[6]/w->bytes);
  w->i=0;
  w->overflow=w->underflow=0;
}

void writeheader(wavfile *w)
{
  LONG l;
  LONG h[11];

  l=w->n*w->bytes;
  h[0]=WAV_RIFF;
  h[1]=l+36;
  h[2]=WAV_WAVE;
  h[3]=WAV_FMT;
  h[4]=16;
  /*   h[5]=WAV_MONO; */
  h[5]=WAV_STEREO;
  h[6]=w->rate;
  h[7]=w->rate*w->bytes;
  h[8]=(w->bits<<16)|w->bytes;
  h[9]=WAV_DATA;
  h[10]=l;
  if(fwrite(h,4,11,w->f)!=11 || ferror(w->f)) errexit(ERR_READ);
  w->i=0;
  w->overflow=w->underflow=0;
}  

void wavstdin(wavfile *w)
{
  w->f=stdin;
  readheader(w);
}

void wavfin(wavfile *w,char *n)
{
  w->f=fileopen(n,"rb",".wav");
  readheader(w);
}

void wavstdout(wavfile *w)
{
  w->f=stdout;
  writeheader(w);
}

void wavfout(wavfile *w,char *n)
{
  w->f=fileopen(n,"wb",".wav");
  writeheader(w);
}

FILE *fileopen(char *n,char *mod,char *ext)
{
  FILE *f;
  char s[128];

  f=fopen(n,mod);
  if(!f)
  {
    if(ext)
    {
      strcpy(s,n);
      strcat(s,ext);
      if(!(f=fopen(s,mod)))
      {
        fprintf(stderr,"%s: Can neither open file %s nor %s\n",PRG,n,s);
        errexit(ERR_PREV);
      };
    } else {
      fprintf(stderr,"%s: Cannot open file %s\n",PRG,n);
      errexit(ERR_PREV);
    };
  };
  openfile[openfiles]=f;
  openfiles++;
  return f;
} 

LONG wavbinget(wavfile *w)
{
  LONG l=0;
  char b;
  LONG i;

  for(i=0; i< w->bytes; i++)
  {
    b=getc(w->f);
    if(b == EOF) errexit(ERR_READ);
    l|=(b&255)<<(8*i);
  };
  w->i++;
  return l;
}

void wavbinput(wavfile *w, LONG l)
{
  LONG v=l;
  LONG b,i;
  
  for(i=0;i<w->bytes;i++)
  {
    b=v&255; v=v>>8;
    if(putc(b,w->f)==EOF) errexit(ERR_WRITE);
  };
  w->i++;
}

VAL wavget(wavfile *w)
{
  VAL x;
  LONG l;
  LONG p = 1<<(w->bytes*8-1);

  l=wavbinget(w);
  if(w->bytes==1)
  {
    x=l/((VAL)p)-1.;
  } else {
    if(l&p)
    {
      l=p-(l^p);
      x=-(VAL)l/((VAL)p);
    } else {
      x=l/((VAL)p);
    };
  };
  return x;
}

void wavput(wavfile *w, VAL x)
{
  /* VAL y; */
  LONG l;
  LONG p = 1<<(w->bytes*8-1);

  if(x>=1.)
  {
    x=MAXAMPL;
    if(!w->overflow)
    {
      w->overflow=1;
      if(verbose) fprintf(stderr,"Overflow: maximum value set.\n");
    };
  }
  else if(x<-1.)
  {
    x=-MAXAMPL;
    if(!w->underflow)
    {
      w->underflow=1;
      if(verbose) fprintf(stderr,"Underflow: minimum value set.\n");
    };
  };
  l=(LONG) floor(w->bytes==1 ? x*p+p : x*p);
  wavbinput(w,l);
}

int getparam(char *s,REAL *v,int msk)
{
  char *p,*q;
  char c;
  REAL x;
  int u = 0;

  x=strtod(s,&p);
  if(p==s) errexit(ERR_PARAM);
  c=*p;
  p++;
  switch(c)
  {
    case 'K': x*=1024.; break;
    case 'k': x*=1000.; break;
    case 'd': x*=0.1  ; break;
    case 'c': x*=0.01 ; break;
    case 'm': x*=0.001; break;
    default : p--;
  };
  q=strstr(" s Hz hz h f B b % rad eg m pt ",p);
  if(!q && *p!=0) errexit(ERR_PARAM);
  switch(*q)
  {
  case ' ': u=U_NU;    break;
    case 's': u=U_SEC;   break;
    case 'H': ;
    case 'h': u=U_HZ;    break;
    case 'f': u=U_HZ;    x*=11025.;    break;
    case 'B': ;
    case 'b': u=U_GAIN;  x=pow(10.,x); break;
    case '%': u=U_GAIN;  x*=0.01;      break;
    case 'r': u=U_PHASE; x/=2*PI;      break;
    case 'e': u=U_PHASE; x/=36.;       break;
    case 'm': u=U_METER; break;
    case 'p': u=U_PT;    break;
    default : errexit(ERR_PARAM);
  };
  if( (u & msk) == 0 ) errexit(ERR_PARAM);
  *v=x;
  return u;
}

REAL gettime(char *s)
{
  REAL t;

  if(getparam(s,&t,U_ALL)!=U_SEC) errexit(ERR_TIME);
  return t;
}

REAL getfreq(char *s)
{
  REAL f;

  if(getparam(s,&f,U_ALL)!=U_HZ) errexit(ERR_TIME);
  return f;
};

int stdopt(char c,char *h)
{
  switch(c)
  {
    case 'v': verbose=1; return 0;
    case 'h': if(h) fprintf(stderr,h,PRG); errexit(0);
    case ':': ;
    case '?': if(h) fprintf(stderr,"Type %s -h for help.\n",PRG);
              errexit(ERR_PREV);
    case 127:
    case -1 : return 1;
  };
  return 0;
};

void setexitmsg(char *prg)
{
  PRG=prg;
};
  
void errexit(int err)
{
  int i;
  for(i=0;i<openfiles;i++) if(openfile[i]) fclose(openfile[i]);
  if(err>ERR_PREV || verbose) fprintf(stderr,"%s: %s.\n",PRG,ERRMSG[err]);
  exit(err);
}


